### ISSUE TITLE
The new Gutenberg editor and the future of ACF

### ISSUE THREAD

**Issue by pagelab** [elliotcondon/acf#622]
Knowing that:

1. The [Gutenberg editor](https://github.com/WordPress/gutenberg) has high probability to be the new WordPress default editor for the 5.0 version,
2. As a result, the current WordPress PHP metabox system [is likely to be greatly limited](https://github.com/WordPress/gutenberg/issues/952#issuecomment-310547291) in favor of a JS flavor of metaboxes. 

My question is: what are the plans for ACF in the near future, as this will change pretty much everything that has been done so far with ACF and affect everyone that works with it.

**Comment by elliotcondon** [comment]
Hi @pagelab 

Thanks for the topic. The Gutenberg UI is very exciting, but I must agree with the majority of developers that the metabox changes are worrying.

I need to do some further reading, but it sounds more like a discussion at the moment as apposed to a definite announcement. 

This is definitely a topic to stay tuned to. If you are following the Gutenberg project, would you mind posting back if there are any major decisions / anouncements made?

Thanks again

**Comment by pagelab** [comment]
> This is definitely a topic to stay tuned to. If you are following the Gutenberg project, would you mind posting back if there are any major decisions / anouncements made?

Sure, I'll post it here as soon as they settle down on the topic of metaboxes. As everybody else, I'm keeping a close eye on this one.

**Comment by marsjaninzmarsa** [comment]
Looks like they plan to support "legacy" metaboxes via PHP callback. It's WordPress after all.

**Comment by pagelab** [comment]
Here's some main Gutenberg tickets concerning the future of PHP metaboxes, FYI:

- Extended section for PHP metaboxes: https://github.com/WordPress/gutenberg/issues/952
- Pluggable areas: https://github.com/WordPress/gutenberg/issues/1352
- Screen options replacement: https://github.com/WordPress/gutenberg/issues/1351

**Comment by elliotcondon** [comment]
Thanks @pagelab - some awesome info here

**Comment by cr101** [comment]
@elliotcondon You will want to read [this comment](https://github.com/WordPress/gutenberg/issues/952#issuecomment-320523428)

**Comment by pagelab** [comment]
- Render "normal" metaboxes created by Advanced Custom Fields: https://github.com/WordPress/gutenberg/pull/2251

**Comment by elliotcondon** [comment]
Thanks @cr101 and @pagelab

**Comment by nylen** [comment]
Please also read https://github.com/WordPress/gutenberg/issues/952#issuecomment-310547291 as it is still very much true.

**Comment by chvon7thal** [comment]
After 30min look to the plugin, I thing we will have to rethink some, but not all, about how to use ACF. For my part, I builded a "block" component, with several type of boxes, beginning  with the notion of columns, able to display different content types, including free formated text, and images.
Probably this part (text and images) won't be necessary anymore. 

I also build a ICS Event component, or a contact description based on vcard, for those there will be the question of where to define/populate them and how to link them... 

For myself I left the idea of custom-post in favor of custom-fields and it gives amazing opportunities for flexibel content and linkings.

But I also short circuited the rendering part of ACF, preferring I kind of caching in the meta-datas: In other words, I use ACF for its flex content system in the admin part. With acf-save I pre-render them. i.e.: For one post I can have one author only, or three (author, translator, photograph) - So if my flex field is populated, I render the format and save it in my own metadata as the author block. The question that I will check: can I create a placeholder for that in Gutenberg and how flexibel would it be (change place, or order with an other guttenberg-custom-block.

**Comment by lukecav** [comment]
A number of comments related to ACF and metaboxes in this post.
https://ma.tt/2017/08/we-called-it-gutenberg-for-a-reason/

**Comment by cr101** [comment]
@elliotcondon Do you plan to release a version of ACF and/or ACF PRO that will be compatible with Gutenberg in WordPress 5.0?

**Comment by elliotcondon** [comment]
Hi guys
Thanks for the comments.

@lukecav - it's hard to believe that WP are even considering removing the WP metaboxes. I am scared for every WooCommerce user.

@cr101 - Yes, of course. Once we get official statements / documentation for the gutenberg changes, I will release updates ASAP to ensure compatibility.

**Comment by lukecav** [comment]
I am sure WooCommerce will not change at all, since those metaboxes are on custom post type for product etc.

**Comment by pagelab** [comment]
Some key points regarding meta boxes from [Matt's post](https://ma.tt/2017/08/we-called-it-gutenberg-for-a-reason/):

> There will definitely be a plugin that gives you the legacy / old edit page.
> There will be no problem to provide a legacy interface for a few releases.

So, is seems that this mitigates a lot the pressure on ACF.

**Comment by lukecav** [comment]
Stress the words for a few releases, again not define major or minor releases of Gutenberg.

**Comment by cr101** [comment]
The Gutenberg development is moving at a very fast pace and the new editor will be WordPress core in only a few months. v1.0 was released today

**Comment by askonomnom** [comment]
I also would love an ACF update that works with Gutenberg considering 1.0 is released and it is rather usable already.

**Comment by elliotcondon** [comment]
Hi guys,

Thanks for the comments. Don't worry, I am keeping up with all Gutenberg news. 

Please don't mistake v1.0 for an official release, the Gutenberg project is far from complete and this is only their versioning style.

Once we get official statements / documentation for the Gutenberg changes, I will release updates ASAP to ensure compatibility. Until then, there is no point worrying.

If anyone is worried about the Gutenberg project direction, please be sure to post on the Gutenberg github issues: https://github.com/WordPress/gutenberg

**Comment by CreativeDive** [comment]
@elliotcondon The answer of the founder and CEO Matt Mullenweg to metaboxes at https://ma.tt/2017/08/we-called-it-gutenberg-for-a-reason/:

> Matt: Some things like toolbar buttons will definitely need to be updated to work with Gutenberg, other things like Metaboxes there will be no problem to provide a **legacy interface for a few releases**. But I would say that plugin authors should start updating their plugins in late September if they want to benefit from Gutenberg’s launch.

He says to metaboxes: ... for a few releases. That means the metaboxes are removed after a few versions, maybe WP 5.3, 5.4 or 5.5. It's very hard to believe this :-( 

But I hope for a full compatibility by ACF and the gutenberg editor and missing meta boxes.

**Comment by brograhamer** [comment]
ok. 

# NO.

ACF is what makes Wordpress a CMS, not just a blogging platform. Blocks cannot, and will never, replace what ACF is. Blocks can, however, enhance custom themes and add some functionality to users without the use of ACF and / or a third party builder.

As a theme developer for a company that builds custom websites using Wordpress, the removal of these metaboxes **is entirely unacceptable** without discussing a solution. 

Why don't we keep metaboxes until we have an alternative that ties in with Gutenberg flawlessly ***before*** we even discuss their removal?

1. How are plugin authors like ACF going to adapt to Gutenberg? Far as I can tell there are **no** real answers here. Also entirely unacceptable. 

2. **I am EXTREMELY excited for Gutenberg, I love it** - but it will literally destroy businesses who rely on Wordpress to build custom websites for their clients. 

I would prefer to never manually create my own metaboxes ever again without the help of ACF.

Yes, I understand a solution is being worked on - but NOBODY here knows what that even looks like or if it's possible. So far this seems pretty heavy handed against businesses and plugin creators with the "figure it out, open source, not out problem" montra. 

I keep commenting here for any sign of a solution. I keep leaving with more questions, more anxiety, and more frustrations. This is absolutely ridiculous. 

# Will metaboxes that ACF and Yoast provides exist under the Gutenberg edit (not legacy)?

This should be dead simple to answer. Sadly, not finding those answers, and my pure excitement and love for how it functions / looks is hinging on this very fact.

**Comment by lukecav** [comment]
@CreativeDive 

Matt Mullenweg is one of the co-founders of WordPress, with the other co-founder being Mike Little.
https://en.wikipedia.org/wiki/Mike_Little

**Comment by CreativeDive** [comment]
@lukecav thanks for this information ;-)

**Comment by brograhamer** [comment]
@elliotcondon It was meant for here, but certainly can be posted there as well. I think I speak for everyone when I say this entire process is extremely exciting but equally as frustrating.

**Comment by pagelab** [comment]
https://ma.tt/2017/09/on-react-and-wordpress/

> (...) the Gutenberg team is going to take a step back and rewrite Gutenberg using a different library. It will likely delay Gutenberg at least a few weeks, and may push the release into next year.

**Comment by brograhamer** [comment]
Yikes. How big of an impact is this on Wordpress itself, for those of us unfamiliar with React?

**Comment by lukecav** [comment]
@brograhamer 

React is not in WP core yet, but seems like it will push WP 5.0 out to 2018. Preact and Vue seem to be the top replacements for React use in Gutenberg.

**Comment by pagelab** [comment]
Metabox support is in the works (tested with ACF):

https://github.com/WordPress/gutenberg/pull/2804

**Comment by elliotcondon** [comment]
Thanks @pagelab - sounds promising

**Comment by joelstransky** [comment]
I just gave Gutenberg a spin. They're off to a great start on a tool with lots of potential but I have to agree with a lot of the reviews in that it's really only an improvement on the_content edits. I was happy to see that there's now an "Extended Settings" section where metaboxes will re-emerge.
This could mean some very cool things for ACF in terms how Field Groups get associated with their location rules. Field groups as Gutenberg block types would be a cool thing to see. I'm actually really excited to see where this leads (and would love to participate in whatever work gets us there)

**Comment by cr101** [comment]
Gutenberg 1.5 adds [Initial Support for Meta Boxes](https://wptavern.com/gutenberg-1-5-adds-initial-support-for-meta-boxes-makes-gutenberg-the-default-editor)
It's encouraging to see progress is made towards custom fields support. I certainly hope that there will be a away to remove/hide the 'post content' editor for CPTs as it takes too much space and it pushes the custom fields too far down the page. 
We rarely use the current tinyMCE editor in our projects.

**Comment by lukecav** [comment]
Not showing metaboxes by default on load, until you click on Extend Settings, seems like a bit of a mess in Gutenberg as well as.
![new product extend settings wc](https://user-images.githubusercontent.com/5797332/32115079-ada92e60-bb02-11e7-8394-3b60e7a3cd7e.png)
https://github.com/WordPress/gutenberg/issues/3165

**Comment by kevinwhoffman** [comment]
Hey ACF friends, I share many of your concerns outlined in this issue, and I've been trying to bring them to light in the Gutenberg repo, most recently in https://github.com/WordPress/gutenberg/issues/3151#issuecomment-339473518. Please voice your concerns there as well.

**Comment by lukecav** [comment]
At least the **Extend Settings** element was removed from Gutenberg, so that metaboxes show on load now.

**Comment by shanejones** [comment]
I've just noticed that Metaboxes are back although I've also noticed that the ACF fields that I have on the page won't save when I update the post. It just changes to this and doesn't really do anything. 

![Screenshot](https://user-images.githubusercontent.com/325424/32322102-3810d758-bfbb-11e7-988a-18975d6d395a.jpg)

**Comment by kevinwhoffman** [comment]
@shanejones Content not saving has been documented here. Please give it a thumbs up or add a comment if you have additional info. https://github.com/WordPress/gutenberg/issues/3277

**Comment by shanejones** [comment]
@kevinwhoffman Weirdly, I demo'd this to someone yesterday and it eventually saved. The version that didn't save was a very complex custom field setup. Fingers crossed for it to get fixed.

**Comment by TimVevida** [comment]
Maybe an interesting feature in Gutenberg 1.17: nested blocks. See https://github.com/WordPress/gutenberg/pull/2743
Also, meta boxes are no longer in an iframe: https://github.com/WordPress/gutenberg/pull/3345

**Comment by lukecav** [comment]
I guess someone really does like the Lego, blocks analogy.

![batman](https://user-images.githubusercontent.com/5797332/32900357-80789914-caaa-11e7-953b-8e97330e09c6.gif)

**Comment by shanejones** [comment]
So I've just updated to 1.7

Seems to save some fields but not others. I have a site that uses flexible content and that doesnt seem to save when the posts saves. 

Essentially the Gutenberg area saves but not the flexible content. When I refresh the page I see this even though the post has saved. 

![screen shot 2017-11-17 at 10 19 44](https://user-images.githubusercontent.com/325424/32942427-f1b522f6-cb80-11e7-9af2-811ebf942458.jpg)

**Comment by rogerburkhard** [comment]
I have the same problem. I see the fields but can't save Repeater Fields with Gutenberg 1.7.

**Comment by pagelab** [comment]
The Gutemberg team are considering the meta box support in Gutemberg "feature complete".

https://github.com/WordPress/gutenberg/issues/952#issuecomment-348008624

**Comment by elliotcondon** [comment]
Hi @pagelab. Thanks for the heads up.
Feature complete..... 🤞

**Comment by lukecav** [comment]
Is that comment for real, it will not be feature complete until it works as expected. Ironic considering Gutenberg is still a beta until it gets merged into WP core in 6-12 months

**Comment by shanejones** [comment]
@pagelab @elliotcondon Just tested that with the development site I was having issues with. Updated Gutenberg and ACF to the latest version and it's still problematic

Changed content of some custom field items (site uses flexible content blocks) on a page. Clicked update on the post. Refreshed the page and still see this
![screen shot 2017-11-30 at 08 46 03](https://user-images.githubusercontent.com/325424/33421870-6f4b7920-d5ab-11e7-811e-fc2bd92c84bb.jpg)

**Comment by NewJenk** [comment]
Hi @elliotcondon,

How's work going on making sure ACF works smoothly with Gutenberg when it launches?

**Comment by elliotcondon** [comment]
Hi @NewJenk 

We are keeping our eyes and ears open for all Gutenberg news whilst also testing each new version they release.

There is not much we can do until WP release some "official" documentation for us. Don't fear, we will make sure that ACF continues to work with Gutenberg.

**Comment by lukecav** [comment]
@NewJenk

A number of ACF related issues have been created on the Gutenberg repo. So I am sure ACF will still work great when Gutenberg is in WP core.

https://github.com/WordPress/gutenberg/issues?utf8=%E2%9C%93&q=acf

**Comment by pagelab** [comment]
There's a [new plugin out there](https://wordpress.org/plugins/gutenberg-custom-fields/) that works in a similar way ACF does – but using Gutemberg APIs. [Here's the blog post](https://riad.blog/2017/12/11/with-gutenberg-what-happens-to-my-custom-fields/) about it.

It's potentially useful as an example in how to migrate ACF itself to the new block paradigm.

**Comment by elliotcondon** [comment]
Thansk @pagelab

**Comment by rianhall** [comment]
That plugin is exactly what I was looking for.

**Comment by llity** [comment]
Hi @elliotcondon hope you can integrate ACF into Gutenberg, in this case, I can use Gutenberg display the custom fields with a custom layout, and fewer code skills to modify the source files. Thanks!

**Comment by iantsch** [comment]
Dear ACF community,
I need ACF to work with the new Gutenberg, so I created a [wrapping plugin](https://github.com/iantsch/acf-gutenberg). It will use ACF and adds a React wrapper for Gutenberg. If there is a field group it will lock the template and render the ACF form. It is a proof of concept, but I will try to push this further if Gutenberg meta box handling will break ACF on release.
I'd like to get some feedback, so feel free to test [acf-gutenberg](https://github.com/iantsch/acf-gutenberg) and create issues.
Thanks!

**Comment by cr101** [comment]
@iantsch It would be great to be able to store the ACF metadata in custom database tables. Storing the ACF metadata in one table (wp_postmeta) is not scalable as it makes searching and reporting painfully slow.

**Comment by elliotcondon** [comment]
@iantsch Great work! I'll be sure to check this out in the new year when I continue in my Gutenberg experiements.

@cr101. Please keep this thread on topic.

Thanks guys.

**Comment by danieltj27** [comment]
Leaving a reply as a bookmark. Glad to see the discussion around this has already started a while back. This is something that has concerned me for a few months regarding how ACF will fit into the new editing experience.

I've seen from the WordPress Slack that Gutenberg is expected to land in April 2018 with the 5.0 release so that will come around shortly for sure.

**Comment by RevConcept** [comment]
Following ;)

**Comment by verteb** [comment]
Following too!

**Comment by cr101** [comment]
Check out this [new Gutenberg developer toolkit](https://github.com/ahmadawais/create-guten-block)

**Comment by nikosolihin** [comment]
Following

**Comment by sokratesagogo** [comment]
Gutenberg version 2.2.0 adds template selection to page attributes. Unfortunately an ACF Location rule based on selection of a given template does not seem to work with this Gutenberg implementation

**Comment by sterndata** [comment]
AARGH:  ACF fields do not save on page creation/update, just as I'm doing a presentation on "It's safe to use ACF and Gutenberg".

**Comment by danieltj27** [comment]
The JavaScript for ACF and Gutenberg have lots of conflicts at the moment.

**Comment by RevConcept** [comment]
@sterndata - that's THE WORST.

**Comment by yellowweb** [comment]
Following

I really hope ACF will be worked with upcoming Wordpress-Update and the new Editor.

**Comment by daviddarke** [comment]
@yellowweb me too 😬

**Comment by sterndata** [comment]
Thanks, Elliot.  5.6.9 fixed my Gutenberg issue.  (My GutenError)  My presentation is back on!

**Comment by shanejones** [comment]
A lot more fields are now saving in the latest version. 

@elliotcondon The only fields that are problematic now are the Flexible Content blocks which do not save at all.

**Comment by leeapogee** [comment]
This situation is a nightmare, I have a hard enough time as it is keeping 70+ sites up and running, the wp 'team' (read isolated bunch of bastards not living in the real world as far as I'm concerned) introducing functionality I don't need/want, my customers don't need/want (and won't have a clue how to use) that also breaks basic functionality of existing sites is ludicrous! As it stands I think it's time to turn off auto update, this could ruin my business trying to fix sites that weren't broken (and I really do not want to turn off auto updates)!

**Comment by leeapogee** [comment]
If you have a hobby site about poly-dactyl cats or some shit on wordpress.com I'm sure it's all wonderful and fluffy, much like the cats.... But do the devs have a clue that they could be about to deal a devastating blow to the biggest CMS platform on the planet??

**Comment by elliotcondon** [comment]
Hi @leeapogee. If you would like your comments heard, they should be directed to the WP Gutenberg team, not here on the ACF repository. Thanks 👍

**Comment by samuel-hauser** [comment]
Thanks für the update; 5.6.9 looks nice. 

Moreover: The compatibility of ACF with Gutenberg seems pretty smooth. Feels almost like the fuss about Automatic's decision to introduce Gutenberg was a bit exaggerated.

Result of my first testing: There seem to be some teething troubles in 5.6.9 though. 
--> Some text fields (WYSIWYG) don't seem editable in compatibility mode.

Gutenberg is activated.
ACF Custom fields and Gutenberg are loaded visually as expected - except for the WYSIWYG-fields:

Webdev console reports that `tinymce.js` could not been found in `wp-includes` folder.

_I guess that's why I can't neither focus nor edit the WYSIWYG fields in my flexible content block and why they look like a normal text field._

**Comment by CreativeDive** [comment]
Will the simultaneous use of Gutenberg as the main editor and TinyMCE as the editor for ACF fields be possible with the release of WordPress 5.0?

**Comment by danieltj27** [comment]
I sense a Guten Field, right @elliotcondon? ;-)

**Comment by gschoppe** [comment]
> Feels almost like the fuss about Automatic's decision to introduce Gutenberg was a bit exaggerated.

@samuel-hauser Don't mistake the incredible and substantial work that @elliotcondon is doing in making ACF compatible with two simultaneous rendering systems that are completely different, for the issue being a triviality.

There are months of effort being spent by plugin developers around the world to support this change, and until 5.0 releases, you won't really see the ones that either didn't manage to have the time/budget or that were fundamentally incompatible.

There was never a question of whether ACF would release a Gutenberg version. That's a testament to his dedication to the community. But the same cannot be said for all of the 40k+ active plugins out there.

**Comment by cgrymala** [comment]
I'll have to do a little more testing to make sure this isn't a compatibility issue, but, based on my preliminary testing of ACF Pro 5.6.9, datetime fields still aren't saving when used in Gutenberg.

In my preliminary testing, at least, I'm using ACF Pro embedded inside of a custom plugin, with the fields defined in PHP. I'm not sure, yet, whether that makes a difference or not. I'll report back after doing some more isolated testing, unless someone else gets a chance to confirm my findings in the meantime. Thanks. 

_WordPress 4.9.4, Gutenberg 2.2.0 and ACF Pro 5.6.9._

**Comment by sterndata** [comment]
Just a bit of update on the WYSIWYG field, as of this moment:

1.  Text mode works, but visual does not.
2.   given a field named 'wysiwyg', with content 'text' entered in text mode
`$field = get_field( 'wysiwyg') ` returns false even thought here's content, but `the_field( 'wysiwyg');` prints the output.

**Comment by lukecav** [comment]
https://plugincompat.danielbachhuber.com/#compatibility-results/view-plugin-details/5a95f3dd4915723fa0b27d87/

**Comment by pkarjala** [comment]
Testing with WP 4.9.4, Gutenberg 2.3.0, ACF Pro 5.6.9

When editing only ACF Fields on a Post, the edits do not notify the Gutenberg editor that changes have been made, so the "Update" button does not highlight.  Not sure how this would wire up.

**Comment by CreativeDive** [comment]
@pkarjala I have the same issue. Only after editing the gutenberg editor the save button is active. But an other issue is, the field data will doesn't save after saving a post.

The gutenberg editor with wordpress 5.0 will comes in april 2018.

**Comment by pkarjala** [comment]
@CreativeDive Which field data do you mean?  The ACF field data?  I'm able to get it to save after editing the main post, and call it using `the_field()` calls.

**Comment by CreativeDive** [comment]
@pkarjala I mean the data of an ACF field. I have installed a clean wordpress with the latest gutenberg editor and after enter some vales in a ACF field. After saving the post, the fields are empty.

**Comment by tomwake** [comment]
I've been following this for a little while and conducting my testing - great work @elliotcondon. I was quite surprised at how seamless it all seemed when I first tried ACF and Gutenberg together.

I've just done a test multisite install with a flexible content WYSIWYG editor and it's currently failing on the visual tab with 404s for -
/wp-content/plugins/gutenberg/vendor/skins/lightgray/content.min.css?wp-mce-4607-20180123  
/wp-content/plugins/gutenberg/vendor/skins/lightgray/skin.min.css?wp-mce-4607-20180123

Has anyone else come across this?

**Comment by thedonquixotic** [comment]
Hey I've been out of the loop for the past couple of weeks moving to a new job and been without internet and stuff. I was curious if the Gutenberg and acf stuff had been fixed? Last I checked I was having compatibility problems and heard in the forums there would be an update soon but was curious to hear more. I am planning on releasing on big announcements on my website and I love writing and editing in Gutenberg but acf is necessary for the proper final presentation of my work so I don't know if I should disable it for now or not.

**Comment by squarecandy** [comment]
I just spent most of today doing some testing of various Gutenberg and ACF situations for my own edification, so I thought I would post my findings on the current state of things here.

I tried out both WordPress 5.0-alpha-42808 (Bleeding edge nightly March 8) and WordPress 4.9.4 with the following plugins:

* Gutenberg 2.3.0
* Classic Editor 0.2
* ACF Pro 5.6.9 

I created [this field group for testing](https://gist.github.com/squarecandy/b9c28dd9ee4cc2a13fa6547202de3d70) that uses (almost) every ACF pro field group and feature/situation including repeaters, tabs, conditionals, and more.


## Supporting Legacy Sites

So seems like a big concern for a lot of people is just making sure your existing sites that make extensive use of ACF don't break when Gutenberg rolls out.

Good news! The Classic Editor plugin will take care of everything for you if you just want to stick with TinyMCE for existing sites.

To test now:

- Download the [Gutenberg Plugin](https://wordpress.org/plugins/gutenberg/), enable it. 
- _"Help! Everything is broken! Also all of my clients over the age of 35 are yelling at me because 'everthing is different'..."_
- Download [Classic Editor](https://wordpress.org/plugins/classic-editor/)
- go to **Settings > Writing** and uncheck the "Do not replace the editor. Add alternate links..." checkbox
- _"Phew!" Back to classic TinyMCE editing with ACF running normally and a still-up-to-date-and-secure WordPress core._


Sounds like the plan is still to release Gutenberg as part of the WordPress 5.0 core so at that point all you'll need to do is use the Classic Editor plugin.

## Supporting Gutenberg for main content area editing while also supporting ACF Pro 5

So the Gutenberg editor is actually pretty cool and I think will be very useful once we can all wrap our brains around it. I think many people will want to start using it soon on new projects while still supporting ACF features. Here's what I found about using Gutenberg with ACF Pro 5.6.9

- Major meta box support freak out concerns from 2017 have been worked out. The concept that the whole WordPress core support for meta boxes and fields that ACF relies on will be tossed out the window with the introduction of Gutenberg is no longer on the table. Thank you to everyone who made a lot of noise about this last year. 
- Generally, I lot of the bugs seem to have been worked out too. Things are getting close to being able to just use Gutenberg as the editor for the_content and ACF for your custom fields as usual.

There are still a few bugs, but they seem like things that can be worked out between now and the official 5.0 release.

### Current ACF related Bugs

There are still a few issues with the dirty form state and ACF or other `meta_box` using plugins. Some of these have been addressed in the past, but still seem to be popping up.

- if you edit an existing page, don't change the title or body and change an ACF field, you can't click the Update button because the form state still thinks it is unchanged. See https://github.com/WordPress/gutenberg/issues/5520
- if you do update a page with ACF fields and then try to leave it gives you a "do you want to leave" warning. See https://github.com/WordPress/gutenberg/issues/4687 (maybe these two are related??)
- Missing stylesheets cause WYSIWYG fields to fail. See https://github.com/WordPress/gutenberg/issues/5522 .  I got this to work again temporarily by just copying `/wp-includes/js/tinymce/skins/*` to `/wp-content/plugins/gutenberg/vendor/skins`

Other than these minor issues, I got every ACF field type to work smoothly underneath a Gutenberg editor today. Awesome!!!

## The Future

Obviously some really smart people are on the case of imagining a bright future where ACF is even more integrated into the Gutenberg blocks system, like some of the things @pagelab mentions earlier in this thread. Hooray for that in the long run - let us know how we can help. In the meantime, it's good to know that things appear to be on track to keep legacy sites running smoothly and also start to integrate Gutenberg into new projects alongside ACF without having to wait for a major rewrite of ACF.

**Comment by pagelab** [comment]
https://www.billerickson.net/getting-your-theme-ready-for-gutenberg/

> In the next six months, I expect ACF (or a plugin like it) to pivot to block building, providing the tools for developers to easily create the blocks they need for custom theme development with minimal time and a unified design.

ACF as a block builder. That's a nice idea to explore in the future. What do you think @elliotcondon ?

**Comment by squarecandy** [comment]
> ACF as a block builder.

I love this idea. However, I think considering support for legacy sites built with ACF is important alongside any ground up reimaginings. Perhaps the 5.x branch becomes an "LTS" version and still receives security updates for a few years and 6.x becomes a ground up rebuild that takes full advantage of Gutenberg?

**Comment by nicooprat** [comment]
Also read an interesting [post about Elementor 2](https://elementor.com/upcoming-elementor-v2/), mainly these parts:

> A brand-new process igniting dynamic content creation that will blow your mind, including an **extensive integration of Advanced Custom Fields** & Toolset

and

> The new version of **Elementor will be completely compatible with Gutenberg**, and we are now considering a variety of ways to connect the new WordPress editor with Elementor. Once it’s out, you will be able to write your post in Gutenberg, then design it with Elementor.

Will be really interesting to see how ACF, Gutenberg and Elementor will work together in the next few months...!

**Comment by gschoppe** [comment]
> I love this idea. However, I think considering support for legacy sites built with ACF is important alongside any ground up reimaginings. Perhaps the 5.x branch becomes an "LTS" version and still receives security updates for a few years and 6.x becomes a ground up rebuild that takes full advantage of Gutenberg?

It's important to remember that Gutenberg doesn't officially sunset the traditional post-edit screen. It simply replaces it for post-types where the Gutenberg editor is enabled (post and page, by default). Matt Mullenweg has made cryptic statements about how long the traditional editor will remain an option, but Mattias Ventura has also commented on the belief that the classic editor will remain long term for certain use cases.

As such, ACF needs to support both forms of registering and rendering metafields for the foreseeable future, preventing any ground-up reimagining, unless it is as a separate plugin or add-on.

**Comment by squarecandy** [comment]
@gschoppe - thanks, good to know. Hopefully we'll get some less cryptic info about the timeline once 5.0 is officially released.

**Comment by tdmalone** [comment]
> ACF as a block builder.

Advanced Custom Blocks ;)

**Comment by rchipka** [comment]
Ideally, ACF fields are converted into reusable Gutenburg block types and then field groups are rendered as block settings for specified block types/locations.

Please check out [this example/proof-of-concept code](https://gist.github.com/rchipka/4c62e8aa450e34218485d8c724c0d83b) for adding ACF fields to the Block Settings panel in Gutenburg.

It adds a location rule for "Gutenburg Block Type", which can be a string like "core/columns".

Once a rule is added, you can click on a block of that type and see your fields appear in the Block Settings panel.

Field data is saved to the block's "attributes", which can be used to render stuff on a custom block (in JS or in PHP).

Once they implement a way to extend existing block types, these fields can also modify the rendering of existing block types to incorporate settings applied in the fields.

I plan to move this code to a plugin until ACF implements a good alternative, but the solution needs to get over a few Gutenburg/React hurdles first.

Please leave a 👍 on [this issue](https://github.com/WordPress/gutenberg/issues/5907) to help get things moving forward.

Sidenote: ACF currently overwrites `window.wp` and `window.wp.hooks`, which breaks things in Gutenburg. The best workaround right now is to just rename it to `window.wp.acf_hooks`.

**Comment by yellowweb** [comment]
Hallo liebe Kunden,

ich gönne mir und MattGelb bis zum 07.05.2018 eine Auszeit. 

In dieser Zeit nehme ich wohlverdienten Urlaub, bilde mich weiter um für Euch auf dem aktuellsten Stand zu bleiben, beschäftige mich mit der DSGVO und schaffe mir Raum für Inspiration und Schaffenskraft. 

Ich lese meine E-Mails unregelmässig, sollte ein dringender Fall vorliegen melde ich mich definitiv.

Ich freue mich auf den 07.05.2018. 
Frisch und voller Tatendrang.

Liebe Grüße und eine wundervolle Zeit,

Katy Schmitt

**Comment by cr101** [comment]
@rchipka  you might want to take a look at [Gutenberg Custom Fields](https://github.com/youknowriad/gcf)

**Comment by rchipka** [comment]
@cr101 Thanks, this is exactly what I'm trying to do.

The only distinction is that instead of undertaking the massive task of rewriting all ACF fields and functionality into Gutenburg React components, I'm attempting to use Gutenburg hooks to inject ACF fields (in all their current glory) into the React component system.

The benefit of this approach is that we can pretty much use ACF fields as they are now without having to wait for someone else to come along and rewrite everything- which takes lots of time, support and effort, not to mention could cause new bugs and old regressions.

**Comment by thedonquixotic** [comment]
Hey, I just wanted to say that I'm super interested in this project @rchipka I'm unfortunately too novice of a dev to help very much though. I work as an interaction designer, though I'm taking frontend classes on the side. If you ever need any beta testers let me know!

**Comment by rchipka** [comment]
ACF Gutenburg - Version 1.0.0
https://github.com/rchipka/wp-acf-gutenburg

Okay everyone, let's test this.

Make your ACF fields appear within certain Gutenburg blocks (or within their settings panel) by targeting the block type in the field group's Location rules.

*Note:* You will need to replace all occurrences of `wp.hooks` with `wp.acf_hooks` within the file `/plugins/advanced-custom-fields-pro/assets/js/acf-input.min.js`

**Comment by taylorbryant** [comment]
@rchipka This is really cool. Great work, man!

**Comment by thedonquixotic** [comment]
Can we also get some screenshots of how it looks @rchipka ? I'm going to try this out on my local build this weekend but I'm busy every night this week and I'm curious. I shared your stuff with the WP ATX slack though.

**Comment by rchipka** [comment]
![screen shot 2018-04-03 at 4 46 49 pm](https://user-images.githubusercontent.com/7538316/38275172-01bda020-375f-11e8-9346-ad6cf3ed6653.png)

## Example Post
![screen shot 2018-04-03 at 4 47 26 pm](https://user-images.githubusercontent.com/7538316/38275175-04a4f57c-375f-11e8-9dfd-8917b9dab715.png)

## Position: "normal"
![screen shot 2018-04-03 at 4 47 39 pm](https://user-images.githubusercontent.com/7538316/38275181-0a420484-375f-11e8-96af-1516224147e5.png)

## Position: "side"
![screen shot 2018-04-03 at 4 48 27 pm](https://user-images.githubusercontent.com/7538316/38275187-0d236436-375f-11e8-80b4-eea7203c271c.png)

**Comment by cr101** [comment]
@rchipka Will your plugin be of any use for people like myself who register the ACF fields via PHP instead of using the UI?

**Comment by JoshFialkoff** [comment]
Is there an ACF-authored plugin that we can use on test sites which use Gutenberg?

**Comment by pkarjala** [comment]
@JoshFialkoff You can install the current version of ACF and it will work with the ACF block underneath the Gutenberg editor.

**Comment by JoshFialkoff** [comment]
I see @pkarjala. I was assuming that ACFs would look more Gutenberg-like. Here's a screenshot.
<img width="821" alt="2018-04-06 14-05-12" src="https://user-images.githubusercontent.com/3923170/38436821-a7e5cf7c-39a3-11e8-931d-f17607b7782c.png">

**Comment by pkarjala** [comment]
That's the ongoing discussion!  😄 This is the current "gap" implementation until the full version update of Gutenberg is released.

**Comment by JoshFialkoff** [comment]
I understand. I'm saying I don't want to have to use a third party solution
@patrick.

**Comment by thedonquixotic** [comment]
@rchipka Tried the plugin just now and got it activated but there seems to be a problem when I actually try to write a post. I've put an issue on the github page for the plugin.

**Comment by rchipka** [comment]
@thedonquixotic This will be fixed in the upcoming version.

I revised the code to better integrate with Gutenberg 2.6.0.

Also, I'm considering renaming it to **Advanced Custom Blocks**, as long as that's cool with @tdmalone.

In addition to major stability/integration improvements, this version will add the ability to make a field group its own block type (with the option to allow child blocks).

The ability to make a field group a block type will allow you to use the Gutenberg add block ("+") to select (by name) the ACF field group and show the fields for that field group in the block content (or in the sidebar).

The latest plugin will also require installing **Gutenberg Block Templates** (which I'll be releasing as a separate plugin).

This plugin will add the ability to easily add or modify the layout for any block type (in PHP) by placing a PHP file with the block's name into your theme's `blocks` directory.

This is especially powerful because it allows composing custom, nestable, PHP-rendered Gutenberg blocks with integrated ACF fields.

And, since it's all done in a PHP block template file, a developer doesn't need to know any JS/React in order to get a simple block with some fields up and running.

Think of it as a WordPress template "partial", but for block content.

Advanced Custom Blocks will hook into this plugin's AJAX call in order to save custom block attributes before rendering the block content/template.

It will also hook into `get_field` as an intuitive alternative for getting an attribute value from within the block template.

**Comment by tdmalone** [comment]
@rchipka Haha no permission from me required, but I am honoured to have provided the inspiration for the name ;)

I wonder though if we should take this discussion to a new issue in your repo, so we don’t clutter this issue for Elliot?

**Comment by tdmalone** [comment]
Glad you're working on it E, but it's a pity it's under wraps! Why not develop alongside the community? :)

**Comment by cjke** [comment]
It is a commercial project - there is probably a level of risk around ownership depending on the level of help he accepts.

**Comment by rchipka** [comment]
@elliotcondon Thanks for the update.

Hopefully Advanced Custom Blocks serves as a temporary solution and perhaps a model for what goes into ACF core.

The same goes for Gutenburg Block Templates. Ideally a similar solution is implemented in Gutenberg core.

**Comment by elliotcondon** [comment]
Hi @tdmalone 

I'm keeping things private until Gutneberg is in a more finished and positive place.
What I can tell you is that our integration is working, and it's exciting to think about all the new opportunities this will open up for ACF developers!

**Comment by rchipka** [comment]
@CreativeDive This is caused by ACF overwriting `window.wp.hooks` object, causing `wp.hooks.hasFilter` to be undefined.

A workaround is to replace all occurrences of `wp.hooks` with `wp.acf_hooks` within the file `/plugins/advanced-custom-fields-pro/assets/js/acf-input.min.js`

**Comment by CreativeDive** [comment]
@rchipka Thanks for helping, but I will wait for the fix by @elliotcondon

**Comment by rchipka** [comment]
@CreativeDive Actually, `wp.hooks` is no longer overwritten in version 5.6.10

Unfortunately, it uses the existing (incompatible) `wp.hooks` functions instead of creating its own `acf_hooks` object.

Now the problem is that the existing Gutenberg JavaScript WP hooks API requires the use of a namespace parameter when interfacing with hooks, so ACF's hook API calls have to be converted over to the new Gutenberg compatible calls.

As long as ACF continues to package its own JS hook/event management system with each release, it should really just put everything into its own `acf_hooks` object until its ready to integrate with the existing API and remove the packaged one.

**Comment by elliotcondon** [comment]
Hi guys

Thanks for the notifications.
Yes, I'm aware of the `wp.hooks` issues with Gutenberg and have opened a ticket with the core team: https://github.com/WordPress/gutenberg/issues/6091

For now, I plan to move the ACF hooks into the `acf` JS object and avoid any conflicting 3rd party libraries.

**Comment by mdgmichael** [comment]
Hi! I'm very curious about whether/how this has progressed—either in @rchipka’s plugin or in ACF core. My team has a library of custom block types that we've built over the years using ACF's Flexible Content fields, and we're eager to convert them into Gutenberg blocks.

My instinct at this moment is that @rchipka has the right fundamental approach: find a way to place an ACF field group in the settings for a block type, and save the values to the block's attributes as JSON. (I tried using his plugins in our sandbox, but they were last updated in April and I don't think they've survived the last few rounds of WP/Gutenberg updates.)

I also know @elliotcondon was working on a parallel track. Has there been any update on details and/or when we might get our hands on it?

Thanks to all from a massive ACF fan.

**Comment by coollorenzo** [comment]
Yeah would love if ACF flexible content would become and easy way to add custom Gutenberg blocks. Just what I need as a webdesigner

**Comment by thedonquixotic** [comment]
Hey, I've been out of the loop on WP stuff because I've been working on some backend projects but I plan on making some huge updates in the next couple of weeks, and would love to give these plugins a whirl. I'll take a look at the Advanced Custom Blocks plugin github myself but hearing about the process from the devs themselves would also be helpful, as I'm still pretty limited in my code skills (but getting better everyday!)

Thank you!

**Comment by samuel-hauser** [comment]
Check the ShopTalk with Elliot Condon: https://t.co/t2toSDaSzY

**Comment by RevConcept** [comment]
@samuel-hauser thanks for sharing that! 

First, off - I am 100% of the same mindset Elliot is with what ACF "is" and how much is should have nothing to do with the Gutenberg updates...that's the biggest thing I don't understand about this whole plan...there are SO MANY developers using WP in a way that has nothing to do with posts and blogs in a traditional sense, and a glorified page builder is not what my client base needs or wants. 

Second, it was such a relief to here that ACF will for sure remain stable and hopefully adapt at some point...but the passion behind it from Elliot is what I liked best! My entire business is built on top of that plugin, ha. That's not to say we wouldn't adapt...but why mess with perfection??

*Also cannot believe that Auttomattic has never approached him, at all?! Maybe they really are that out of touch with what WP devs are doing...

**Comment by eliot-akira** [comment]
Here's a link to directly listen to the relevant portion of the ShopTalk podcast:

http://shoptalkshow.com/episodes/323-acf-elliot-condon/#t=39:57

**Comment by mdgmichael** [comment]
Whoah! Ask and ye shall receive: https://www.advancedcustomfields.com/blog/the-state-of-acf-in-a-gutenberg-world/

This is _exactly_ what I was hoping for, and we barely even have to wait for it. I'm legit excited to start using ACF Blocks. Thank you, @elliotcondon! This ACF Pro customer is very happy!

**Comment by justiny** [comment]
Perhaps I'm missing something here, but I'm not seeing the location > rules > field group option for "Block".  Is there a function that needs added for that option to show?  
![acf-gutenberg-blocks-2](https://user-images.githubusercontent.com/2721886/44045521-fdc799dc-9ee5-11e8-83d1-eaf9c974d504.png)

**Comment by elliotcondon** [comment]
Hi Guys.

The upcoming "ACF Blocks" feature is in it's infancy and has not yet been included in the public version. Stay tuned to our blog for updates on this new feature! We plan to have this out in the next month or two :)

**Comment by enoversum** [comment]
@elliotcondon Will there be or is there some kind of early-access-beta-try-before-you-die for the blocks feature? Eagerly awaiting its arrival!

**Comment by craigpearson** [comment]
It'd be really beneficial to get early access to this if possible. Mainly because it sounds exciting, but more importantly so we can start planning in any needed refactors and what they may look like across installations. At the moment we're leaning towards embracing Gutenberg alone for layouts, but it'd be nice to show the team that we don't have to

**Comment by TwoPaperdolls** [comment]
You might want to leave a note on the blog post that block isn't available in the location rules yet.

I spent a while chasing my tail trying to figure out what I was doing wrong until I finally came here to see what was up.  I had enabled early access and it wasn't available for me either.

**Comment by adamr415** [comment]
Yes, the blog post is a bit miselading since there is even a code example included. I too jumped right in then scratched my head a while.

**Comment by rchipka** [comment]
For those who want to start using something like this right away, you may want to check out the latest version of [rchipka/advanced-custom-blocks](https://github.com/rchipka/advanced-custom-blocks).

## Add a block and target location

![screen shot 2018-08-23 at 7 23 20 pm](https://user-images.githubusercontent.com/7538316/44556941-25c55100-a70a-11e8-8ff5-b51c312be386.png)

Block location rules work for Block Name as well as others (like Post Type).


## Set block name and icon

![screen shot 2018-08-23 at 7 23 33 pm](https://user-images.githubusercontent.com/7538316/44556962-49889700-a70a-11e8-89de-b8c6f9693cd8.png)

Block Name is defined at the bottom of the Field Group settings page as well as an option custom Block Icon (using dashicons slug).

## Select our block

![screen shot 2018-08-23 at 7 13 59 pm](https://user-images.githubusercontent.com/7538316/44556981-6755fc00-a70a-11e8-9a69-47a9aad03f3d.png)

We can select our block under Widgets by default. As you can see, our block takes on the Field Group name and the custom icon.

## Post edit screen

![screen shot 2018-08-23 at 7 21 38 pm](https://user-images.githubusercontent.com/7538316/44557063-d4699180-a70a-11e8-8f55-3b8528852361.png)

Our block takes on all the Field Group fields and layout. Here we have two Testimonial blocks inside of a Columns block.

Note that we're able to seamlessly add (and render) multiple instances of the same fields on the same post type (even though it's not a repeater field!). This allows the user to create their own custom repeater-like layout.

## Render your block

Note that in order for our block to become available in the Gutenberg editor, we never had to write any code to actually register that block.

This implementation does not require any custom block registration mechanism ahead of time. Instead it simply creates the block and uses a filter for rendering that block's content.

### Filters:
```add_filter('acf/render_block', function ($output, $attributes) {```
```add_filter('acf/render_block/name=block_name', function ($output, $attributes) {```

```php
<?php
add_filter('acf/render_block/name=testimonial', function ($output, $attributes) {
  ob_start();
  ?>
  <div class="testimonial">
    <img src="<? the_field('avatar'); ?>" />
    <h3><? the_field('author'); ?></h3>
    <span><? the_field('testimonial'); ?></span>
  </div>
  <?
  return ob_get_clean();
}, 10, 2);
```

Note: For even simpler block templating, an extension could be created that hooks `render_block` and attempts to auto-load a block template from within a directory like `blocks/block-name.php`.

**Comment by thedonquixotic** [comment]
Love it. Going to try this this week. I just finished rebuilding my work computer and just need to get my server connections back up :) Woo!

**Comment by thedonquixotic** [comment]
Hey, got Robbie's plugin installed and working, though still working through some possible limitations of use it has for my needs. 

That said, and this doesn't do the exact same thing, I thought I should link this other plugin that I found while searching stuff that some people here might find useful: https://github.com/youknowriad/gcf

**Comment by thedonquixotic** [comment]
Oooh, actually I just found a plugin that does exactly what I need it to do. Technically it doesn't hook into Advanced Custom Field and it only has text and text area field types but it does what I need it to do and I like the UI a bit more than Robbie's. Maybe it's something that Robbie or Elliott would want to contribute to it or perhaps would find the code useful to their own projects: https://github.com/rheinardkorf/advanced-custom-blocks

It's also got a killer logo, I gotta say.

**Comment by adamr415** [comment]
@rchipka is your project what will be implemented into ACF. I want to get a jump on transitioning my ACF sites but wasn't sure if it was worth the effort until it was finalized into ACF.

**Comment by rchipka** [comment]
@adamr415 No, it's a completely separate implementation.

That being said, anything you build using my version can be easily ported over to the ACF implementation based on the examples provided in the blog post.

It would only require the adding `acf_register_block` for each block as well as updating the location rules to use ACF's "Block Name" rule.

**Comment by dillix** [comment]
Is there any news on `acf_register_block` implemetation in ACF 5/ACF 5 PRO?

**Comment by battmates** [comment]
> Is there any news on `acf_register_block` implemetation in ACF 5/ACF 5 PRO?

I had this back from ACF via email a couple of days ago:

> I am afraid that the "ACF Blocks" feature that was showcased is in early beta, and is not yet ready to be released. We will be blogging about this feature over the coming months 🙂 Hopefully it will be available soon. Unfortunately, at this time we do now have any ETA on when the feature may be available.

**Comment by dillix** [comment]
My ACF 5 PRO don't work via metabox also when Gutenberg enabled.

**Comment by rchipka** [comment]
## v2.1.1

Done:

* ~~Built-in block template targeting~~ (`blocks/acf-block-name.php` or `blocks/acf/block-name.php`)
* ~~Store ACF field data in block attributes (in addition to post_meta)~~
* ~~Allow customizing the block's group (right now it's in Widgets)~~

Changed:

 * Store one or more of a field's values in post_meta (in a more queryable way)
 * Use [actions for render hooks](https://github.com/rchipka/advanced-custom-blocks/issues/8#issuecomment-427095243) (instead of filters)

Fixed:

* Getting/storing multiple blocks with same field on the same post

Todo:

* Add block location rule restrictions based on parent's "Block Name"
* Register new field type "Gutenberg Blocks" so that we can nest blocks as fields
* Enable sidebar position for field group
* Add block count restrictions

**Comment by maxhart7** [comment]
> Hi Guys.
> 
> The upcoming "ACF Blocks" feature is in it's infancy and has not yet been included in the public version. Stay tuned to our blog for updates on this new feature! We plan to have this out in the next month or two :)

Hi @elliotcondon ! Any update on when we can expect this? Or is there a way Pro members can get early access? Eager to start testing this and familiarising myself and my colleagues on how to use it before 5.0 launches.

Thanks!

**Comment by rchipka** [comment]
@maxhart7 If you'd like to get an early start, you can try [rchipka/advanced-custom-blocks](https://github.com/rchipka/advanced-custom-blocks), which actually requires fewer steps to implement than the solution proposed in the original ACF blog post.

The main distinction at this point is ACF's implementation requires `acf_register_block`, whereas Advanced Custom Blocks handles this for you (since it's completely redundant anyways).

In fact, I see no reason why the eventual ACF release wouldn't adopt an API similar to the one implemented in my version.

It's a pretty straightforward concept, the challenge for ACF is getting a product-ready version released after significant testing.

If you'd like to be a part of that testing, try the open source version and report any issues you run into.

Contributions are always welcome and don't forget to :star: if you like it.

**Comment by rchipka** [comment]
Looks like ACF v5.8.0-beta1 contains the functionality we've been waiting for.

Seems the implementation is the same as mentioned in the original blog post, with `acf_register_block`, etc.

**Comment by rchipka** [comment]
There's currently a bug where if the Gutenberg plugin is loaded after ACF, then ACF Blocks won't be available.

Add this workaround to your functions.php file:
```
if ( !defined('GUTENBERG_VERSION') ) {
  define( 'GUTENBERG_VERSION', '4.1.1' );
}
```

**Comment by cr101** [comment]
@rchipka the Gutenberg version changes frequently. would that create any issues?

**Comment by rchipka** [comment]
@cr101 If you're referring to my fix with the hardcoded version, then no it shouldn't cause any issues since ACF doesn't use the version number for anything.

This (in theory) won't be a problem once Gutenberg is included in Wordpress core.
